testnet command and make localnet-start

greg-szabo opened the pull request:
The purpose of this change is to introduce a `testnet` command, similar to the one in Tendermint. The command creates separate folders with initialized configuration. This enables easy testnet setup and configuration management.

The `testnet` command is used when creating configuration for local testnets using docker. One such example is created under the `localnet-start` target in the Makefile.

<!-- Thanks for filing a PR! Before hitting the button, please check the following items.-->

* [ ] Updated all relevant documentation in docs
* [X] Updated all code comments where relevant
* [ ] Wrote tests
* [X] Updated CHANGELOG.md

rigelrozanski commented on cmd/gaia/app/genesis.go:
```diff
@@ -86,18 +86,21 @@ type GaiaGenTx struct {
 
 // Generate a gaia genesis transaction with flags
-func GaiaAppGenTx(cdc *wire.Codec, pk crypto.PubKey) (
+func GaiaAppGenTx(cdc *wire.Codec, pk crypto.PubKey, config *gaiacfg.Config) (
```
A design pattern we're trying to follow in the SDk: let's not make `config` a pointer here, if we want to modify config, let's return a new copy of it.

rigelrozanski commented on config/config.go:
```diff
@@ -0,0 +1,45 @@
+	Name		string			`mapstructure:"name"`
+	CliRoot	string			`mapstructure:"homeclient"`
+	Init		*InitConfig		`mapstructure:"init"`
+	GenTx		*GenTxConfig	`mapstructure:"gentx"`
```
what's up with this formatting?

rigelrozanski replied:
looks like go-fmt is not enabled? Bunch of stuff is misformatted within this file

rigelrozanski commented on server/init.go:
```diff
@@ -118,6 +80,54 @@ func GenTxCmd(ctx *Context, cdc *wire.Codec, appInit AppInit) *cobra.Command {
 	return cmd
 }
 
+func gentxWithConfig(config *cfg.Config, gaiaConfig *gaiacfg.Config, ctx *Context, cdc *wire.Codec, appInit AppInit) (
```
convention, `ctx` should always be the first input to a function if it exists

rigelrozanski replied:
also let's not use pointers for the config's if possible

greg-szabo replied:
I fixed the `ctx` convention, but in the Tendermint code, every function that handles the config is expecting a pointer. If I modify it here, my code will be full with &config horribleness.

For example, a few lines below, I'll have to change this:
```
	pubKey := readOrCreatePrivValidator(config)
```
to this:
```
	pubKey := readOrCreatePrivValidator(&config)
```

rigelrozanski commented on server/testnet.go:
```diff
@@ -0,0 +1,240 @@
+
+	gaiad testnet --v 4 --o ./output --populate-persistent-peers --starting-ip-address 192.168.10.2
+	`,
+		RunE: func(_ *cobra.Command, _ []string) error {
```
RunE should not define the function in this situation - it's just too large and it get's confusing with all the added indentation. Define a separate function and take it as an argument. Ex:

```golang 
type inp struct {
     cdc *wire.Codec, 
     appInit AppInit,
}

func TestnetFilesCmd(ctx *Context, cdc *wire.Codec, appInit AppInit) *cobra.Command {
     inp := inp{cdc, appInit}

     cmd := &cobra.Command{
		Use:     "testnet",
		Short:  "Initialize files for a Gaiad testnet",
		RunE:   inputs.gregsFunc, 
}

func (c inp)gregsFunc(_ *cobra.Command, _ []string) error {
     something that uses c.cdc
     something else that uses c.appInit
}
```

rigelrozanski commented on server/testnet.go:
```diff
@@ -0,0 +1,240 @@
+
+				err := os.MkdirAll(filepath.Join(nodeDir, "config"), nodeDirPerm)
+				if err != nil {
+					_ = os.RemoveAll(outputDir)
```
why are we skipping this input? (for below RemoveAll as well)

greg-szabo replied:
It's a "best-effort" cleanup (copied from the Tendermint code). We already have an error message, so while we're quitting with that message, we're trying to remove the half-baked configuration folders that we created. If it succeeds, good, if not, we still want to present the higher priority error message to the user. (And let the user do the cleanup.)

rigelrozanski commented on server/testnet.go:
```diff
@@ -0,0 +1,240 @@
+	populatePersistentPeers bool
+	hostnamePrefix          string
+	startingIPAddress       string
+	p2pPort                 int
```
let's not use any global variables whatsoever here - let's pass variables around as necessary

rigelrozanski commented on server/testnet.go:
```diff
@@ -0,0 +1,240 @@
+)
+
+func init() {
+}
```
don't need to define an empty init (can just remove this function)

rigelrozanski commented on server/testnet.go:
```diff
@@ -0,0 +1,240 @@
+
+const (
+	nodeDirPerm = 0755
+)
```
const nodeDirPerm = 0755

rigelrozanski commented on server/testnet.go:
```diff
@@ -0,0 +1,240 @@
+
+			// Generate private key, node ID, initial transaction
+			for i := 0; i < nValidators; i++ {
+				nodeDirName := cmn.Fmt("%s%d", nodeDirPrefix, i)
```
we're moving away from using `cmn.Fmt` just use `fmt.Sprintf` directly

rigelrozanski commented on server/testnet.go:
```diff
@@ -0,0 +1,240 @@
+			return nil
+			},
+	}
+	cmd.Flags().IntVar(&nValidators, "v", 4,
```
let's stick with the convention in the rest of the sdk for now and just use `viper.GetString` instead of cobra "var" flags

rigelrozanski commented on server/testnet.go:
```diff
@@ -0,0 +1,240 @@
+}
+
+func hostnameOrIP(i int) string {
+	if startingIPAddress != "" {
```
rather than indenting the entire function for a single if, just flip it and return at the beginning:

```
hostname := fmt.Sprintf("%s%d", hostnamePrefix, i)
if startingIPAddress == "" {
    return hostname
}
... the other code
...
return hostname
```

rigelrozanski commented on config/config.go:
```diff
@@ -0,0 +1,45 @@
+	ChainID		string			`mapstructure:"chainid"`
+	GenTxs		bool			`mapstructure:"gentxs"`
+	GenTxsDir	string			`mapstructure:"gentxsdir"`
+	Overwrite	bool			`mapstructure:"overwrite"`
```
I'm confused as to how `config` is uses - maybe add a comment at the top of this file... also what is `mapstucture` used for?

greg-szabo replied:
Completely thrown out as discussed.

rigelrozanski commented:
Looks cool, I know this is still WIP - but I added some interim comments anyways... keep at it!

rigelrozanski commented on server/init.go:
```diff
@@ -18,6 +18,7 @@ import (
 	"github.com/tendermint/go-crypto/keys"
 	"github.com/tendermint/go-crypto/keys/words"
 	cfg "github.com/tendermint/tendermint/config"
+	gaiacfg "github.com/cosmos/cosmos-sdk/config"
```
import should be in organized with other cosmos imports

rigelrozanski commented on config/config.go:
```diff
@@ -0,0 +1,45 @@
+package config
```
After reviewing in more detail - it's pretty clear to me that `config.go` actually just belongs in the `server/` package - it's configuration options for the server commands nothing else (not client, nor keys, etc.). There is no import problems - it's only referenced in  `server` and `gaia` right now - so after the move, it will only be imported by `gaia`

greg-szabo commented:
Fixed all the above and finished the `gaiad testnet` command. Also introduced the `make localnet-start` `make localnet-stop` commands. Please review from scratch, lots changed.

greg-szabo commented:
Making the GenTxConfig struct (formerly GaiaConfig) part of the server package creates an interesting loop:
1. server package defines GenTxConfig struct
2. **server.AppInit.AppGenTx is a function that has a GenTxConfig type input among other input (to hand over configuration)**
3. mock package defines mock AppGenTx function - result: mock package requires the server package imported so GenTxConfig parameter can be defined 
4. server package tests the Init command. It uses the mock package's AppGenTx function for mocking an initial transaction. - result: server package requires the mock package

If the parameter handed over wasn't a struct, but a primitive, this loop wouldn't have happened. I think the resolution is to keep GenTxConfig separately.

rigelrozanski commented on docker-compose.yml:
```diff
@@ -0,0 +1,68 @@
+      - ./build:/gaiad:Z
+    networks:
+      localnet:
+        ipv4_address: 192.168.10.2
```
is this file supposed to be ignored in this PR? has custom IPs

greg-szabo replied:
The file is integral part of the Makefile changes, it's not supposed to be ignored. This file generates a local testnet with 4 nodes on an internal docker network. The custom IPs are static IPs for the nodes, only used within the docker container. There's no impact on your network outside of it.

rigelrozanski commented on server/testnet.go:
```diff
@@ -0,0 +1,174 @@
+	ipv4 := net.ParseIP(ip).To4()
+	if ipv4 == nil {
+		fmt.Printf("%v: non ipv4 address\n", ip)
+		os.Exit(1)
```
Should pass be passing errors instead of just exiting here

greg-szabo replied:
Yep, agreed. This was also copied from Tendermint as-is. I'll change it to pass an error instead.

rigelrozanski commented on server/testnet.go:
```diff
@@ -0,0 +1,174 @@
+
+		err = os.MkdirAll(clientDir, nodeDirPerm)
+		if err != nil {
+			_ = os.RemoveAll(viper.GetString(outputDir))
```
we should add `viper.GetString(outputDir)` to a variable instead of calling it so many times

greg-szabo replied:
Ok, I'm on it.

rigelrozanski commented on server/testnet.go:
```diff
@@ -0,0 +1,174 @@
+	}
+	cmd.Flags().Int(nValidators, 4,
+		"Number of validators to initialize the testnet with")
+	cmd.Flags().String("o", "./mytestnet",
```
I think you meant to use `outputDir` instead of `"o"`

greg-szabo replied:
Interestingly, this is the input parameter in Tendermint and I wanted to keep the two compatible. I can get rid of that idea, since the code changed so much. I'm renaming it to `--outputdir`.

rigelrozanski commented on server/testnet.go:
```diff
@@ -0,0 +1,174 @@
+	cmd.Flags().String(nodeDirPrefix, "node",
+		"Prefix the directory name for each node with (node results in node0, node1, ...)")
+
+	cmd.Flags().String(startingIPAddress, "192.168.0.1",
```
who's IP is the default?

greg-szabo replied:
The default is a random IP I chose because it's non-routable. We can change it to whatever we want.

rigelrozanski commented on server/init.go:
```diff
@@ -207,6 +200,68 @@ func InitCmd(ctx *Context, cdc *wire.Codec, appInit AppInit) *cobra.Command {
 	return cmd
 }
 
+func initWithConfig(ctx *Context, cdc *wire.Codec, appInit AppInit, config *cfg.Config, gaiaInitConfig InitConfig) (
```
nothing should use the naming "gaia" outside of the gaia example directory - should rename `gaiaInitConfig` to something more generic that's fitting for it's usage

greg-szabo replied:
initConfig it is.

rigelrozanski commented on server/init.go:
```diff
@@ -118,6 +112,46 @@ func GenTxCmd(ctx *Context, cdc *wire.Codec, appInit AppInit) *cobra.Command {
+	if err != nil {
+		return
+	}
+	err = cmn.WriteFile(file, bz, 0644)
```
is there a common Ensure AND Write call in common?

greg-szabo replied:
Nope, let's discuss.

greg-szabo replied:
A quick glimpse at tmlibs yielded no results for an ensure + write function. :(

greg-szabo commented on cmd/gaia/app/genesis.go:
```diff
@@ -70,9 +69,16 @@ func GaiaAppInit() server.AppInit {
+		FlagName:       flagName,
+		FlagClientHome: flagClientHome,
+		FlagOWK:        flagOWK,
+	}
```
I thought about that a lot, here's the reason behind it:
 I want to make sure we only define flag names in one place. This used to be the server package, `init.go`. The problem is that now, because of the testnet command, I need to hand over the parameters through AppInit which is in turn used in the app package (genesis.go). In the genesis.go file I don't have access to the flag names if they are set as global variables. If they are passed over like this, I can query viper for their value.

greg-szabo commented:
Successfully messed up my rebase. After learning how to do it properly, the changes went over to this PR: https://github.com/cosmos/cosmos-sdk/pull/1208
